File: python/journal/field_remap.py
"""Rust-compatible field-name remapping for high-level directory writers."""

from __future__ import annotations

import hashlib


REMAPPING_MARKER = 'ND_REMAPPING'
REMAPPED_PREFIX = 'ND_'
MASK64 = (1 << 64) - 1

LOWERCASE = 0
UPPERCASE = 1
DOT = 2
UNDERSCORE = 3
HYPHEN = 4

TOKEN_LOWERCASE = 0
TOKEN_UPPERCASE = 1
TOKEN_CAPITALIZED = 2

FIELD_LOWERCASE = 0
FIELD_UPPERCASE = 1
FIELD_LOWER_CAMEL = 2
FIELD_UPPER_CAMEL = 3
FIELD_EMPTY = 4

SEP_DOT = 0
SEP_HYPHEN = 1
SEP_UNDERSCORE = 2


def is_systemd_compatible_log_field_name(name):
    if not isinstance(name, str) or len(name) == 0 or len(name) > 64:
        return False
    first = ord(name[0])
    if first < 0x41 or first > 0x5A:
        return False
    for ch in name:
        c = ord(ch)
        if c == 0x5F or 0x41 <= c <= 0x5A or 0x30 <= c <= 0x39:
            continue
        return False
    return True


def remap_fields(fields, registry):
    out = []
    mappings = []
    pending = {}

    for field in fields:
        name = field['name']
        if is_systemd_compatible_log_field_name(name):
            out.append(field)
            continue

        mapped = registry.get(name) or pending.get(name)
        if mapped is None:
            mapped = encode_remapped_field_name(name)
            pending[name] = mapped
            mappings.append({'original': name, 'mapped': mapped})

        remapped = dict(field)
        remapped['name'] = mapped
        out.append(remapped)

    return out, mappings


def encode_remapped_field_name(field_name):
    if isinstance(field_name, bytes):
        raw = field_name
    else:
        raw = str(field_name).encode('utf-8')

    encoded = _rdp_encode(raw)
    if encoded is None:
        return _md5_fallback(raw)

    if _has_checksum(encoded):
        compressed = encoded[:2] + _compress_runs(encoded[2:])
    else:
        compressed = _compress_runs(encoded)

    normalized = raw.decode('utf-8').upper().replace('.', '_').replace('-', '_')
    if normalized.startswith('RESOURCE_ATTRIBUTES_'):
        normalized = 'RA_' + normalized[len('RESOURCE_ATTRIBUTES_'):]
    elif normalized.startswith('LOG_ATTRIBUTES_'):
        normalized = 'LA_' + normalized[len('LOG_ATTRIBUTES_'):]
    elif normalized.startswith('LOG_BODY_'):
        normalized = 'LB_' + normalized[len('LOG_BODY_'):]

    result = 'ND' + compressed.upper() + '_' + normalized
    if len(result) > 64:
        return _md5_fallback(raw)
    return result


def _md5_fallback(raw):
    return REMAPPED_PREFIX + hashlib.md5(raw).hexdigest().upper()


def _rdp_encode(raw):
    tokens = _tokenize(raw)
    if tokens is None:
        return None
    return _encode_nodes(raw.decode('utf-8'), _parse_tokens(tokens))


def _char_kind(c):
    if 0x61 <= c <= 0x7A:
        return LOWERCASE
    if 0x41 <= c <= 0x5A or 0x30 <= c <= 0x39:
        return UPPERCASE
    if c == 0x2E:
        return DOT
    if c == 0x5F:
        return UNDERSCORE
    if c == 0x2D:
        return HYPHEN
    return None


def _tokenize(raw):
    if len(raw) == 0:
        return []

    tokens = []
    start = 0
    previous = None
    first = None
    has_lowercase = False
    has_uppercase = False

    for i, byte in enumerate(raw):
        current = _char_kind(byte)
        if current is None:
            return None
        if previous is None:
            first = current
            previous = current
            continue

        if previous in (DOT, UNDERSCORE, HYPHEN):
            should_split = True
        elif current in (DOT, UNDERSCORE, HYPHEN):
            should_split = True
        elif previous == UPPERCASE and current == UPPERCASE:
            next_kind = _char_kind(raw[i + 1]) if i + 1 < len(raw) else None
            should_split = next_kind == LOWERCASE
        elif previous == LOWERCASE and current == LOWERCASE:
            should_split = False
        elif previous == UPPERCASE and current == LOWERCASE:
            should_split = has_uppercase and has_lowercase
        else:
            should_split = True

        if should_split:
            tokens.append(_create_token(first, has_lowercase, has_uppercase, start, i))
            start = i
            first = current
            has_lowercase = False
            has_uppercase = False
        else:
            if current == LOWERCASE:
                has_lowercase = True
            elif current == UPPERCASE:
                has_uppercase = True
        previous = current

    if start < len(raw):
        tokens.append(_create_token(first, has_lowercase, has_uppercase, start, len(raw)))
    return tokens


def _create_token(first, has_lowercase, _has_uppercase, start, end):
    if first == LOWERCASE:
        return {'word': True, 'kind': TOKEN_LOWERCASE, 'start': start, 'end': end}
    if first == UPPERCASE:
        return {
            'word': True,
            'kind': TOKEN_CAPITALIZED if has_lowercase else TOKEN_UPPERCASE,
            'start': start,
            'end': end,
        }
    if first == DOT:
        return {'word': False, 'sep': SEP_DOT}
    if first == HYPHEN:
        return {'word': False, 'sep': SEP_HYPHEN}
    return {'word': False, 'sep': SEP_UNDERSCORE}


def _parse_tokens(tokens):
    nodes = []
    if tokens and not tokens[0]['word']:
        nodes.append({'field': True, 'type': FIELD_EMPTY})

    builder = None
    for i, token in enumerate(tokens):
        if not token['word']:
            if builder is not None:
                nodes.append({'field': True, 'type': builder['type']})
                builder = None
            nodes.append({'field': False, 'sep': token['sep']})
            if i + 1 >= len(tokens) or not tokens[i + 1]['word']:
                nodes.append({'field': True, 'type': FIELD_EMPTY})
            continue

        if builder is not None:
            if _can_add(builder['type'], token['kind']):
                builder['extended'] = True
                continue
            if (
                builder['type'] == FIELD_LOWERCASE and
                not builder['extended'] and
                token['kind'] == TOKEN_CAPITALIZED
            ):
                builder['type'] = FIELD_LOWER_CAMEL
                builder['extended'] = True
                continue
            nodes.append({'field': True, 'type': builder['type']})

        builder = {'type': _field_type_for_token(token['kind']), 'extended': False}

    if builder is not None:
        nodes.append({'field': True, 'type': builder['type']})
    return nodes


def _can_add(field_type, token_type):
    return (
        (field_type == FIELD_LOWERCASE and token_type == TOKEN_LOWERCASE) or
        (field_type == FIELD_UPPERCASE and token_type == TOKEN_UPPERCASE) or
        (field_type == FIELD_LOWER_CAMEL and token_type == TOKEN_CAPITALIZED) or
        (field_type == FIELD_UPPER_CAMEL and token_type == TOKEN_CAPITALIZED)
    )


def _field_type_for_token(token_type):
    if token_type == TOKEN_LOWERCASE:
        return FIELD_LOWERCASE
    if token_type == TOKEN_UPPERCASE:
        return FIELD_UPPERCASE
    return FIELD_UPPER_CAMEL


def _encode_nodes(source, nodes):
    has_camel = any(
        node['field'] and node['type'] in (FIELD_LOWER_CAMEL, FIELD_UPPER_CAMEL)
        for node in nodes
    )
    out = _checksum(source) if has_camel else ''

    i = 0
    while i < len(nodes):
        node = nodes[i]
        if not node['field']:
            i += 1
            continue

        next_is_separator = i + 1 < len(nodes) and not nodes[i + 1]['field']
        next_is_field = i + 1 < len(nodes) and nodes[i + 1]['field']
        sep = nodes[i + 1]['sep'] if next_is_separator else SEP_DOT
        out += chr(_pair_char(node['type'], next_is_separator, next_is_field, sep))
        i += 2 if next_is_separator else 1

    return out


def _pair_char(field_type, next_is_separator, next_is_field, sep):
    base = ord('a')
    if field_type == FIELD_LOWER_CAMEL:
        base = ord('f')
    elif field_type == FIELD_UPPER_CAMEL:
        base = ord('k')
    elif field_type == FIELD_UPPERCASE:
        base = ord('p')
    elif field_type == FIELD_EMPTY:
        base = ord('u')

    if next_is_separator:
        if sep == SEP_DOT:
            return base
        if sep == SEP_UNDERSCORE:
            return base + 1
        return base + 2
    if next_is_field and field_type != FIELD_EMPTY:
        return base + 3
    return base + 4


def _checksum(source):
    raw = source.encode('utf-8') + b'\xff'
    value = _sip_hash_13_zero(raw)
    first = (value // 36) % 36
    second = value % 36
    return chr(_checksum_char(first)) + chr(_checksum_char(second))


def _checksum_char(index):
    return ord('A') + index if index < 26 else ord('0') + index - 26


def _has_checksum(encoded):
    return bool(encoded) and ('A' <= encoded[0] <= 'Z' or '0' <= encoded[0] <= '9')


def _compress_runs(value):
    out = []
    i = 0
    while i < len(value):
        ch = value[i]
        count = 1
        while i + count < len(value) and value[i + count] == ch:
            count += 1
        if count <= 2:
            out.append(ch * count)
        else:
            remaining = count
            while remaining > 0:
                if remaining > 9:
                    out.append('9' + ch)
                    remaining -= 9
                elif remaining > 2:
                    out.append(str(remaining) + ch)
                    remaining = 0
                else:
                    out.append(ch * remaining)
                    remaining = 0
        i += count
    return ''.join(out)


def _sip_hash_13_zero(msg):
    v0 = 0x736F6D6570736575
    v1 = 0x646F72616E646F6D
    v2 = 0x6C7967656E657261
    v3 = 0x7465646279746573

    def sip_round():
        nonlocal v0, v1, v2, v3
        v0 = (v0 + v1) & MASK64
        v1 = _rotl64(v1, 13) ^ v0
        v0 = _rotl64(v0, 32)
        v2 = (v2 + v3) & MASK64
        v3 = _rotl64(v3, 16) ^ v2
        v0 = (v0 + v3) & MASK64
        v3 = _rotl64(v3, 21) ^ v0
        v2 = (v2 + v1) & MASK64
        v1 = _rotl64(v1, 17) ^ v2
        v2 = _rotl64(v2, 32)

    offset = 0
    while offset + 8 <= len(msg):
        m = int.from_bytes(msg[offset:offset + 8], 'little')
        v3 ^= m
        sip_round()
        v0 ^= m
        offset += 8

    b = len(msg) << 56
    for i, byte in enumerate(msg[offset:]):
        b |= byte << (8 * i)

    v3 ^= b
    sip_round()
    v0 ^= b
    v2 ^= 0xFF
    sip_round()
    sip_round()
    sip_round()

    return (v0 ^ v1 ^ v2 ^ v3) & MASK64


def _rotl64(value, bits):
    return ((value << bits) | (value >> (64 - bits))) & MASK64

File: go/journal/field_remap.go
package journal

import (
	"crypto/md5"
	"encoding/binary"
	"encoding/hex"
	"math/bits"
	"strings"
)

const (
	remappingMarker = "ND_REMAPPING"
	remappedPrefix  = "ND_"
)

type remappedFieldMapping struct {
	original string
	mapped   string
}

type rdpCharType uint8

const (
	rdpLowercase rdpCharType = iota
	rdpUppercase
	rdpDot
	rdpUnderscore
	rdpHyphen
)

type rdpTokenType uint8

const (
	rdpTokenLowercase rdpTokenType = iota
	rdpTokenUppercase
	rdpTokenCapitalized
)

type rdpSeparator uint8

const (
	rdpSepDot rdpSeparator = iota
	rdpSepHyphen
	rdpSepUnderscore
)

type rdpToken struct {
	word  bool
	kind  rdpTokenType
	start int
	end   int
	sep   rdpSeparator
}

type rdpFieldType uint8

const (
	rdpFieldLowercase rdpFieldType = iota
	rdpFieldUppercase
	rdpFieldLowerCamel
	rdpFieldUpperCamel
	rdpFieldEmpty
)

type rdpNode struct {
	field bool
	ftype rdpFieldType
	sep   rdpSeparator
}

type rdpFieldBuilder struct {
	ftype    rdpFieldType
	start    int
	end      int
	extended bool
}

func isSystemdCompatibleLogFieldName(name string) bool {
	if len(name) == 0 || len(name) > 64 {
		return false
	}
	if name[0] < 'A' || name[0] > 'Z' {
		return false
	}
	for i := 0; i < len(name); i++ {
		c := name[i]
		if c == '_' || (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9') {
			continue
		}
		return false
	}
	return true
}

func remapLogFields(fields []Field, registry map[string]string) ([]Field, []remappedFieldMapping) {
	out := make([]Field, 0, len(fields))
	newMappings := make([]remappedFieldMapping, 0)
	pending := make(map[string]string)

	for _, field := range fields {
		if isSystemdCompatibleLogFieldName(field.Name) {
			out = append(out, field)
			continue
		}

		mapped, ok := registry[field.Name]
		if !ok {
			mapped, ok = pending[field.Name]
		}
		if !ok {
			mapped = encodeRemappedFieldName([]byte(field.Name))
			pending[field.Name] = mapped
			newMappings = append(newMappings, remappedFieldMapping{
				original: field.Name,
				mapped:   mapped,
			})
		}
		out = append(out, Field{Name: mapped, Value: field.Value})
	}

	return out, newMappings
}

func encodeRemappedFieldName(fieldName []byte) string {
	encoded, ok := rdpEncode(fieldName)
	if !ok {
		return rdpMD5Fallback(fieldName)
	}

	var compressed string
	if rdpHasChecksum(encoded) {
		compressed = encoded[:2] + rdpCompressRuns(encoded[2:])
	} else {
		compressed = rdpCompressRuns(encoded)
	}

	normalized := strings.ToUpper(string(fieldName))
	normalized = strings.NewReplacer(".", "_", "-", "_").Replace(normalized)
	if suffix, ok := strings.CutPrefix(normalized, "RESOURCE_ATTRIBUTES_"); ok {
		normalized = "RA_" + suffix
	} else if suffix, ok := strings.CutPrefix(normalized, "LOG_ATTRIBUTES_"); ok {
		normalized = "LA_" + suffix
	} else if suffix, ok := strings.CutPrefix(normalized, "LOG_BODY_"); ok {
		normalized = "LB_" + suffix
	}

	result := "ND" + strings.ToUpper(compressed) + "_" + normalized
	if len(result) > 64 {
		return rdpMD5Fallback(fieldName)
	}
	return result
}

func rdpMD5Fallback(fieldName []byte) string {
	sum := md5.Sum(fieldName)
	return remappedPrefix + strings.ToUpper(hex.EncodeToString(sum[:]))
}

func rdpEncode(fieldName []byte) (string, bool) {
	tokens, ok := rdpTokenize(fieldName)
	if !ok {
		return "", false
	}
	nodes := rdpParse(tokens)
	return rdpEncodeNodes(string(fieldName), nodes), true
}

func rdpCharKind(c byte) (rdpCharType, bool) {
	switch {
	case c >= 'a' && c <= 'z':
		return rdpLowercase, true
	case (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9'):
		return rdpUppercase, true
	case c == '.':
		return rdpDot, true
	case c == '_':
		return rdpUnderscore, true
	case c == '-':
		return rdpHyphen, true
	default:
		return 0, false
	}
}

func rdpTokenize(fieldName []byte) ([]rdpToken, bool) {
	if len(fieldName) == 0 {
		return nil, true
	}

	tokens := make([]rdpToken, 0)
	start := 0
	var prev rdpCharType
	var first rdpCharType
	hasPrev := false
	hasLowercase := false
	hasUppercase := false

	for i := 0; i < len(fieldName); i++ {
		curr, ok := rdpCharKind(fieldName[i])
		if !ok {
			return nil, false
		}
		if !hasPrev {
			first = curr
			prev = curr
			hasPrev = true
			continue
		}

		shouldSplit := false
		switch {
		case prev == rdpDot || prev == rdpUnderscore || prev == rdpHyphen:
			shouldSplit = true
		case curr == rdpDot || curr == rdpUnderscore || curr == rdpHyphen:
			shouldSplit = true
		case prev == rdpUppercase && curr == rdpUppercase:
			if i+1 < len(fieldName) {
				next, ok := rdpCharKind(fieldName[i+1])
				if !ok {
					return nil, false
				}
				shouldSplit = next == rdpLowercase
			}
		case prev == rdpLowercase && curr == rdpLowercase:
			shouldSplit = false
		case prev == rdpUppercase && curr == rdpLowercase:
			shouldSplit = hasUppercase && hasLowercase
		default:
			shouldSplit = true
		}

		if shouldSplit {
			tokens = append(tokens, rdpCreateToken(first, hasLowercase, hasUppercase, start, i))
			start = i
			first = curr
			hasLowercase = false
			hasUppercase = false
		} else {
			if curr == rdpLowercase {
				hasLowercase = true
			} else if curr == rdpUppercase {
				hasUppercase = true
			}
		}
		prev = curr
	}

	if start < len(fieldName) {
		tokens = append(tokens, rdpCreateToken(first, hasLowercase, hasUppercase, start, len(fieldName)))
	}
	return tokens, true
}

func rdpCreateToken(first rdpCharType, hasLowercase, hasUppercase bool, start, end int) rdpToken {
	switch first {
	case rdpLowercase:
		return rdpToken{word: true, kind: rdpTokenLowercase, start: start, end: end}
	case rdpUppercase:
		if hasLowercase {
			return rdpToken{word: true, kind: rdpTokenCapitalized, start: start, end: end}
		}
		return rdpToken{word: true, kind: rdpTokenUppercase, start: start, end: end}
	case rdpDot:
		return rdpToken{sep: rdpSepDot}
	case rdpHyphen:
		return rdpToken{sep: rdpSepHyphen}
	default:
		return rdpToken{sep: rdpSepUnderscore}
	}
}

func rdpParse(tokens []rdpToken) []rdpNode {
	nodes := make([]rdpNode, 0)
	if len(tokens) > 0 && !tokens[0].word {
		nodes = append(nodes, rdpNode{field: true, ftype: rdpFieldEmpty})
	}

	var builder *rdpFieldBuilder
	for i, token := range tokens {
		if !token.word {
			if builder != nil {
				nodes = append(nodes, rdpNode{field: true, ftype: builder.ftype})
				builder = nil
			}
			nodes = append(nodes, rdpNode{sep: token.sep})
			if i+1 >= len(tokens) || !tokens[i+1].word {
				nodes = append(nodes, rdpNode{field: true, ftype: rdpFieldEmpty})
			}
			continue
		}

		if builder != nil {
			if builder.canAdd(token.kind) {
				builder.end = token.end
				builder.extended = true
				continue
			}
			if builder.ftype == rdpFieldLowercase && !builder.extended && token.kind == rdpTokenCapitalized {
				builder.ftype = rdpFieldLowerCamel
				builder.end = token.end
				builder.extended = true
				continue
			}
			nodes = append(nodes, rdpNode{field: true, ftype: builder.ftype})
		}
		builder = &rdpFieldBuilder{ftype: rdpFieldTypeForToken(token.kind), start: token.start, end: token.end}
	}

	if builder != nil {
		nodes = append(nodes, rdpNode{field: true, ftype: builder.ftype})
	}
	return nodes
}

func (b rdpFieldBuilder) canAdd(kind rdpTokenType) bool {
	return (b.ftype == rdpFieldLowercase && kind == rdpTokenLowercase) ||
		(b.ftype == rdpFieldUppercase && kind == rdpTokenUppercase) ||
		(b.ftype == rdpFieldLowerCamel && kind == rdpTokenCapitalized) ||
		(b.ftype == rdpFieldUpperCamel && kind == rdpTokenCapitalized)
}

func rdpFieldTypeForToken(kind rdpTokenType) rdpFieldType {
	switch kind {
	case rdpTokenLowercase:
		return rdpFieldLowercase
	case rdpTokenUppercase:
		return rdpFieldUppercase
	default:
		return rdpFieldUpperCamel
	}
}

func rdpEncodeNodes(source string, nodes []rdpNode) string {
	hasCamel := false
	for _, node := range nodes {
		if node.field && (node.ftype == rdpFieldLowerCamel || node.ftype == rdpFieldUpperCamel) {
			hasCamel = true
			break
		}
	}

	var out strings.Builder
	if hasCamel {
		out.WriteString(rdpChecksum(source))
	}

	for i := 0; i < len(nodes); {
		node := nodes[i]
		if !node.field {
			i++
			continue
		}
		nextIsSeparator := i+1 < len(nodes) && !nodes[i+1].field
		nextIsField := i+1 < len(nodes) && nodes[i+1].field
		var sep rdpSeparator
		if nextIsSeparator {
			sep = nodes[i+1].sep
		}
		out.WriteByte(rdpPairChar(node.ftype, nextIsSeparator, nextIsField, sep))
		if nextIsSeparator {
			i += 2
		} else {
			i++
		}
	}
	return out.String()
}

func rdpPairChar(ftype rdpFieldType, nextIsSeparator, nextIsField bool, sep rdpSeparator) byte {
	base := byte('a')
	switch ftype {
	case rdpFieldLowercase:
		base = 'a'
	case rdpFieldLowerCamel:
		base = 'f'
	case rdpFieldUpperCamel:
		base = 'k'
	case rdpFieldUppercase:
		base = 'p'
	case rdpFieldEmpty:
		base = 'u'
	}
	if nextIsSeparator {
		switch sep {
		case rdpSepDot:
			return base
		case rdpSepUnderscore:
			return base + 1
		default:
			return base + 2
		}
	}
	if nextIsField && ftype != rdpFieldEmpty {
		return base + 3
	}
	return base + 4
}

func rdpChecksum(source string) string {
	msg := append([]byte(source), 0xff)
	hash := sipHash13Zero(msg)
	first := int((hash / 36) % 36)
	second := int(hash % 36)
	return string([]byte{rdpChecksumChar(first), rdpChecksumChar(second)})
}

func rdpChecksumChar(idx int) byte {
	if idx < 26 {
		return byte('A' + idx)
	}
	return byte('0' + idx - 26)
}

func rdpHasChecksum(encoded string) bool {
	if encoded == "" {
		return false
	}
	c := encoded[0]
	return (c >= 'A' && c <= 'Z') || (c >= '0' && c <= '9')
}

func rdpCompressRuns(s string) string {
	if s == "" {
		return ""
	}
	var out strings.Builder
	for i := 0; i < len(s); {
		ch := s[i]
		count := 1
		for i+count < len(s) && s[i+count] == ch {
			count++
		}
		if count <= 2 {
			for j := 0; j < count; j++ {
				out.WriteByte(ch)
			}
		} else {
			remaining := count
			for remaining > 0 {
				if remaining > 9 {
					out.WriteByte('9')
					out.WriteByte(ch)
					remaining -= 9
				} else if remaining > 2 {
					out.WriteByte(byte('0' + remaining))
					out.WriteByte(ch)
					remaining = 0
				} else {
					for j := 0; j < remaining; j++ {
						out.WriteByte(ch)
					}
					remaining = 0
				}
			}
		}
		i += count
	}
	return out.String()
}

func sipHash13Zero(msg []byte) uint64 {
	var k0, k1 uint64
	v0 := uint64(0x736f6d6570736575) ^ k0
	v1 := uint64(0x646f72616e646f6d) ^ k1
	v2 := uint64(0x6c7967656e657261) ^ k0
	v3 := uint64(0x7465646279746573) ^ k1

	remaining := msg
	for len(remaining) >= 8 {
		m := binary.LittleEndian.Uint64(remaining[:8])
		v3 ^= m
		v0, v1, v2, v3 = sipHash13Round(v0, v1, v2, v3)
		v0 ^= m
		remaining = remaining[8:]
	}

	b := uint64(len(msg)) << 56
	for i, c := range remaining {
		b |= uint64(c) << (8 * uint(i))
	}

	v3 ^= b
	v0, v1, v2, v3 = sipHash13Round(v0, v1, v2, v3)
	v0 ^= b
	v2 ^= 0xff
	for i := 0; i < 3; i++ {
		v0, v1, v2, v3 = sipHash13Round(v0, v1, v2, v3)
	}

	return v0 ^ v1 ^ v2 ^ v3
}

func sipHash13Round(v0, v1, v2, v3 uint64) (uint64, uint64, uint64, uint64) {
	v0 += v1
	v1 = bits.RotateLeft64(v1, 13)
	v1 ^= v0
	v0 = bits.RotateLeft64(v0, 32)
	v2 += v3
	v3 = bits.RotateLeft64(v3, 16)
	v3 ^= v2
	v0 += v3
	v3 = bits.RotateLeft64(v3, 21)
	v3 ^= v0
	v2 += v1
	v1 = bits.RotateLeft64(v1, 17)
	v1 ^= v2
	v2 = bits.RotateLeft64(v2, 32)
	return v0, v1, v2, v3
}
